Adjust to match current TF-M config

daverodgman opened the pull request:
## Description

Bring Mbed TLS in line with recent changes to TF-M config

https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/log/lib/ext/mbedcrypto/mbedcrypto_config

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not in 2.28
- [x] **tests** not required

daverodgman commented:
@yanrayw #8164 unsets MBEDTLS_NO_PLATFORM_ENTROPY, which AFAICT is needed to get this to build. Do you know why that was done? Can we remove that change and leave it set, so that it builds cleanly?

daverodgman commented:
> @daverodgman This was done precisely to make the configuration buildable. TF-M disables our entropy function (which is only implemented for Unix and Windows) and provides the function `mbedtls_hardware_poll`. When we run tests based on the TF-M config, since we run them on Linux, we keep `MBEDTLS_NO_PLATFORM_ENTROPY` so as to keep the Linux entropy source in the build.

OK, this makes sense. Thanks.

daverodgman commented on tests/scripts/all.sh:
```diff
@@ -3217,6 +3209,8 @@ component_test_tfm_config_p256m_driver_accel_ec () {
     not grep mbedtls_dhm_ library/dhm.o
     not grep mbedtls_mpi_ library/bignum.o
+    # Check that p256m was built
+    grep -q p256_ecdsa_verify library/libmbedcrypto.a
```
I'm not 100% sure, but I seem to remember that I tried this and it didn't work - so I had to use `p256_ecdsa_verify`. So I'll leave this, unless you've tested it and found otherwise?

daverodgman commented on configs/ext/tfm_mbedcrypto_config_profile_medium.h:
```diff
@@ -80,44 +80,6 @@
- *
- */
-#define MBEDTLS_AES_SETKEY_DEC_ALT
-#define MBEDTLS_AES_DECRYPT_ALT
```
Yes, we no longer need to undef these in `configs/config-tfm.h` as they are no longer set by the TF-M config files.

daverodgman commented:
I've fetched the latest versions of the TF-M files, so assuming no mistakes, at this point in time we should be fully aligned.

mpg reviewed (approved):
LGTM, thanks!

mpg commented:
Ah, this seems to have conflicts with #8568 which is currently in the merge queue. This will need rebasing once #8568 is merged.

daverodgman commented:
I've updated significantly and rebase onto #8568.

The main change is that the files in `ext` are direct copies of the upstream files (modulo license changes, and commenting out a couple of includes referencing files from TF-M). All local changes have been moved to our `config-tfm.h` wrapper. I've updated our tests to refer to this wrapper.

This should significantly simplify updating these files in future.

daverodgman commented:
> @daverodgman should we copy license for `crypto_config_profile_medium.h` as well?

Yes, thanks

daverodgman commented:
Alternatively, we could put something like this in our `config-tfm.h` wrapper, so that it works out of the box when compiling for either baremetal or Linux?

```
#if defined(__IAR_SYSTEMS_ICC__) || defined(__ARM_EABI__)
#define MBEDTLS_NO_PLATFORM_ENTROPY
#endif
```

mpg commented on configs/config-tfm.h:
```diff
@@ -28,3 +28,41 @@
+ * Local changes to crypto config below this delimiter
+ **********************************************************************/
+
+/* Between Mbed TLS 3.4 and 3.5, the PSA_WANT_KEY_TYPE_RSA_KEY_PAIR macro
```
I don't see the point of keeping this commented out this is a about 3.4 -> 3.5 but TF-M has picked 3.5 already, so it's no longer useful. (I'm not sure if it was useful in the first place: RSA and DH are not part of the medium profile, but perhaps still part of the full profile?)

mpg commented on configs/ext/tfm_mbedcrypto_config_profile_medium.h:
```diff
@@ -614,14 +515,55 @@
 
 #if !defined(CRYPTO_HW_ACCELERATOR) && defined(MBEDTLS_ENTROPY_NV_SEED)
-#include "mbedtls_entropy_nv_seed_config.h"
+//#include "mbedtls_entropy_nv_seed_config.h"
```
Random thought: if we wanted to keep this file an exact copy of TF-M's, in order to make future update simpler, we could cheat by defining `CRYPTO_HW_ACCELERATOR` before we `#include` this file from `config-tfm.h`.

daverodgman replied:
Unfortunately, they have:

```
#if !defined(CRYPTO_HW_ACCELERATOR) && defined(MBEDTLS_ENTROPY_NV_SEED)
//#include "mbedtls_entropy_nv_seed_config.h"
#endif

#ifdef CRYPTO_HW_ACCELERATOR
#include "mbedtls_accelerator_config.h"
#endif
```

so we end up with an include we don't want, whether we set it or not.

I am thinking about a follow-up where we get TF-M to gate all the includes by a special variable like `MBEDTLS_NO_INCLUDE`, and maybe also ask them to add the dual license to this file (or simply add a README next to the files that explains that they are dual-licensed even though the header only mentions Apache). Then we can simply do a copy of the TF-M files. @adeaarm would these two changes be possible?

mpg replied:
Ah, good point, I missed that!

adeaarm replied:
Yes, no problem with that addition. Do we really need to re-license this though on TF-M's repo? I don't think there is a particular issue with having the same header files being part of two different repos with two different licenses (which I believe has the same effect of dual licensing the same header anyway). But please let me know if I am missing anything here.

daverodgman replied:
It's simply that ideally we would like to make a copy of the TF-M file with no changes; but it's also desirable to be able to say "all Mbed TLS files (except Everest) are under the dual license". I don't mind making a note in a readme though, if you'd rather not relicense the files.

daverodgman replied:
@adeaarm will you raise an issue to add a macro to disable the includes on your side? If you can do that, I'll make a follow-up PR on our side to copy the file without changes, and add a note in an adjacent file to explain the license situation. Thanks!

mpg reviewed (changes requested):
Almost happy, just two nitpicks and one suggestion that you should feel free to ignore if you don't like it!

mpg reviewed (approved):
LGTM
